### PR TITLE
Libraries: Use reinterpret_cast and singletons

### PR DESCRIPTION
**Pull request by patrickelectric**
This solve some warning messages when using last clang version

### CODE REVIEW

**Comment by WickedShell** [comment]
Why is this needed/what is the value? Just looks like noise to me.

**Comment by patrickelectric** [comment]
This solves a serie of warning messages:
```cpp
warning: cast from 'AP_Param *' to 'SITL::SITL *' increases required alignment from 1 to 8 [-Wcast-align
```
clang 6.0.0

**Comment by jaxxzer** [comment]
LGTM. Merge or close?

**Comment by OXINARF** [comment]
I'd prefer to use the singleton methods, wherever it's possible (and most, if not all, are). SITL is a singleton in master now.

**Comment by patrickelectric** [comment]
@OXINARF thank you, using SITL singleton.

File: libraries/SITL/SIM_ADSB.cpp [hunk 1]
```diff
@@ -73,7 +73,7 @@ void ADSB_Vehicle::update(float delta_t)
 void ADSB::update(void)
 {
     if (_sitl == nullptr) {
-        _sitl = (SITL *)AP_Param::find_object("SIM_");
+        _sitl = reinterpret_cast<SITL *>(AP_Param::find_object("SIM_"));
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply libraries/SITL/SIM_ADSB.cpp]: Splitted and squashed in 1417bae6d5f036222228f469e0a1ba6929c87445

File: libraries/AP_HAL_SITL/SITL_State.cpp [hunk 2]
```diff
@@ -76,12 +76,12 @@ void SITL_State::_sitl_setup(const char *home_str)
     fprintf(stdout, "Starting SITL input\n");
 
     // find the barometer object if it exists
-    _sitl = (SITL::SITL *)AP_Param::find_object("SIM_");
-    _barometer = (AP_Baro *)AP_Param::find_object("GND_");
-    _ins = (AP_InertialSensor *)AP_Param::find_object("INS_");
-    _compass = (Compass *)AP_Param::find_object("COMPASS_");
+    _sitl = AP::sitl();
+    _barometer = reinterpret_cast<AP_Baro *>(AP_Param::find_object("GND_"));
```
Comment (tridge) [inline libraries/AP_HAL_SITL/SITL_State.cpp]: baro, ins and compass can also use singleton methods

**Comment by patrickelectric** [comment]
Thank you for the feedback @tridge, corrections done.

**Comment by magicrub** [comment]
LGTM, merge?

File: libraries/AP_HAL_SITL/SITL_State.cpp [hunk 3]
```diff
@@ -76,10 +76,10 @@ void SITL_State::_sitl_setup(const char *home_str)
...
 
     // find the barometer object if it exists
-    _sitl = (SITL::SITL *)AP_Param::find_object("SIM_");
-    _barometer = (AP_Baro *)AP_Param::find_object("GND_");
-    _ins = (AP_InertialSensor *)AP_Param::find_object("INS_");
-    _compass = (Compass *)AP_Param::find_object("COMPASS_");
+    _sitl = AP::sitl();
+    _barometer = AP::baro().get_instance();
+    _ins = AP::ins().get_instance();
+    _compass = AP::compass().get_singleton();
```
Comment (OXINARF) [inline libraries/AP_HAL_SITL/SITL_State.cpp]: Last three should be in form `AP_Baro::get_instance()`.

**Comment by peterbarker** [comment]
On Mon, 18 Jun 2018, Francisco Ferreira wrote:

> -    _ins = (AP_InertialSensor *)AP_Param::find_object("INS_");
> -    _compass = (Compass *)AP_Param::find_object("COMPASS_");
> +    _sitl = AP::sitl();
> +    _barometer = AP::baro().get_instance();
> +    _ins = AP::ins().get_instance();
> +    _compass = AP::compass().get_singleton();
> 
> Last three should be in form AP_Baro::get_instance().


Not just AP::baro(), AP::ins() and AP::compass()?

**Comment by OXINARF** [comment]
> Not just AP::baro(), AP::ins() and AP::compass()?

@peterbarker Can be, but that demands more changes. Actually looks like some aren't even needed anymore.

**Comment by patrickelectric** [comment]
@peterbarker @OXINARF I agree that we need a singleton standard nomenclature for pointer access in singletons. I'll update baro to use `get_instance`.

**Comment by OXINARF** [comment]
@patrickelectric 

I think @WickedShell actually prefers get_singleton because instance is commonly used in libraries for the drivers.

In any case, that's not the change I was saying that is needed. You are using `AP_Baro::baro().get_instance()` and that translates to `(*AP_Baro::get_instance()).get_instance()` which obviously doesn't make sense. You should just use `AP_Baro::get_instance()` directly. Same applies to INS and Compass.

**Comment by patrickelectric** [comment]
@OXINARF done. Should I rename only `Compass` to `get_singleton` or all others (baro, ins) ?

**Comment by patrickelectric** [comment]
@OXINARF Ty, but maybe in another day :)
Corrections done.

**Comment by patrickelectric** [comment]
@OXINARF any update ?

**Comment by OXINARF** [comment]
@patrickelectric Sorry, for some reason I was convinced I had merged it. Merged, thanks!

**Pull request merged**